radixdlt/src/main/java/com/radixdlt/consensus/liveness/WeightedRotatingLeaders.java

```diff
@@ -0,0 +1,142 @@
+/*
+ * (C) Copyright 2020 Radix DLT Ltd
+ *
+ * Radix DLT Ltd licenses this file to you under the Apache License,
+ * Version 2.0 (the "License"); you may not use this file except in
+ * compliance with the License.  You may obtain a copy of the
+ * License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied.  See the License for the specific
+ * language governing permissions and limitations under the License.
+ */
+
+package com.radixdlt.consensus.liveness;
+
+import com.radixdlt.consensus.View;
+import com.radixdlt.consensus.validators.Validator;
+import com.radixdlt.consensus.validators.ValidatorSet;
+import com.radixdlt.crypto.ECPublicKey;
+import com.radixdlt.utils.UInt128;
+import com.radixdlt.utils.UInt256;
+import java.util.Comparator;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Map.Entry;
+
+/**
+ * Rotates leaders with those having more power being proposed more often
+ * in proportion to the amount of power they have.
+ *
+ * Calculation of the next leader is dependent on the weight state of the
+ * previous view and thus computing the leader for an arbitrary view can
+ * be quite expensive.
+ *
+ * We resolve this by keeping a cache of some given size of the previous
+ * views closest to the highest view calculated.
+ *
+ * This class stateful and is NOT thread-safe.
+ */
+public final class WeightedRotatingLeaders implements ProposerElection {
+
+	private final ValidatorSet validatorSet;
+	private final Comparator<Entry<Validator, UInt256>> weightsComparator;
+	private final CachingNextLeaderComputer nextLeaderComputer;
+
+	public WeightedRotatingLeaders(ValidatorSet validatorSet, Comparator<Validator> comparator, int cacheSize) {
```

Comment:
Not sure what the purpose of the comparator is here, given that the only real option is the validator power, and the power already has a comparator?

Reply:
It's possible for power and weights to be equal and need a deterministic way of choosing the winner

Reply:
Moved to #79